build/gen_stub.php

```diff
@@ -1813,47 +1835,63 @@ protected function pName_FullyQualified(Name\FullyQualified $node) {
 
 function funcInfoToCode(FuncInfo $funcInfo): string {
     $code = '';
-    $returnType = $funcInfo->return->type;
+    $nativeReturnType = $funcInfo->return->type;
+    $phpDocReturnType = $funcInfo->isMethod() ? $funcInfo->return->phpDocType : null;
+    $returnType = $nativeReturnType ?? $phpDocReturnType;
```

Comment:
nit: Out of scope and easy to add later, but it would be useful to have a file doc comment flag (and/or function comment) to disable this, such as `@no-generate-tentative-return-type`, for PECLs that will support php versions older than 8.1

- The extension maintainers might not plan to change the real return type in the near future until the minimum PECL version is 8.1
- The extension maintainers may want the reflection info to be the same in <= 8.0 and >= 8.1
- The phpdoc might just be for IDE/readability purposes and/or be unconfirmed

Reply:
Yeah, that's a good idea, and most probably php-src itself would also need it (see `PDOException::getCode()`). Unfortunately, the RFC is failing hard currently :/ I'm not exactly sure if it's only due to the reflection part or people see fundamental problems with adding return types to internal methods :(

BTW: How is phan affected by the reflection-related changes? Would you also like the solution suggested by Marco, so that `ReflectionMethod::getReturnType()` would return these tentative types?

Reply:
> Unfortunately, the RFC is failing hard currently :/ I'm not exactly sure if it's only due to the reflection part or people see fundamental problems with adding return types to internal methods :(

I'd voted for it - It seemed like this was the most pragmatic way to migrate to real return types incrementally. I can understand that there's some libraries that extend php functionality and deliberately return different classes with similar method signatures, but right now old class methods would have no return type and new class methods would almost all have return types.

So without adding return types to reflection, phan and other static analysis tools would be less effective at reporting definite redundant/impossible conditions such as type assertions and null/false checks (due to the possibility that a user-defined subclass somewhere overrid the type, or the type signatures bundled with the analyzer were inaccurate (e.g. many internal methods return false on failure, but that largely changed in 8.0))

> BTW: How is phan affected by the reflection-related changes? Would you also like the solution suggested by Marco, so that `ReflectionMethod::getReturnType()` would return these tentative types?

I prefer the current approach of adding hasTentativeReturnType and getTentativeReturnType

- If getReturnType was to be modified instead, older versions of Phan would falsely report that the missing return type from user-defined subclasses would be a fatal error at runtime (rather than a deprecation notice). Unit testing libraries that haven't been updated might generate stricter mocks than they need to (prior to 9.0), etc.
- If it passed, it may eventually be possible to have a getReturnType of BaseType, mixed, etc and a getTentativeReturnType of SubClassType, `never`, etc

Reply:
Thank you, @TysonAndre for making an IMO good case for the current implementation in your mail replies! I've just added support for your proposed `@no-generate-tentative-return-type`